### PR TITLE
Update README

### PR DESCRIPTION
**Pull request by rashmigottipati**
Summary of changes:
- Update README with definitions of statistics, and recommended sliding window length and sliding factor. 
- Explained sliding window concept

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -36,21 +36,47 @@ $ make
 ```
```
Comment (kindermoumoute) [inline README.md]: Line 19: add OS X.

Comment (kindermoumoute) [inline README.md]: Line 24: pre-built binary should be found in [plugin releases](https://github.com/intelsdi-x/snap-plugin-processor-statistics/releases) or in [plugin catalog](https://github.com/intelsdi-x/snap/blob/master/docs/PLUGIN_CATALOG.md#processors)? @nanliu

File: README.md [hunk 2]
```diff
@@ -36,21 +36,47 @@ $ make
...
 This builds the plugin in `/build/rootfs`
```
Comment (kindermoumoute) [inline README.md]: This builds the plugin in `/build/<OS>/<ARCH>`.

File: README.md [hunk 3]
```diff
@@ -36,21 +36,47 @@ $ make
...
 
+Unit testing:
+```
+$ make test-small
+```
+
 ### Configuration and Usage
 * Set up the [snap framework](https://github.com/intelsdi-x/snap/blob/master/README.md#getting-started)
 * Ensure `$SNAP_PATH` is exported
```
Comment (kindermoumoute) [inline README.md]: Remove everything that talk about $SNAP_PATH.

File: README.md [hunk 4]
```diff
@@ -36,21 +36,47 @@ $ make
...
+17) 2nd-percentile - value below which 2% observations may be found 
+18) 9th-percentile - value below which 9% observations may be found 
+19) 25th-percentile - value below which 25% observations may be found 
+20) 75th-percentile - value below which 75% observations may be found 
+21) 91st-percentile - value below which 91% observations may be found  
+22) 95th-percentile - value below which 95% observations may be found 
+23) 98th-percentile - value below which 98% observations may be found 
+24) 99th-percenticle - value below which 99% observations may be found 
+
+To illustrate the sliding window concept, let's say we have a sliding window of 10 over 100 points long, if the sliding factor is 1 and we want to calculate the statistic "average", it will take the average from 0-10, 1-11, 2-12, 3-13 and so on. That takes the 100 data points and averages to 99 points.
```
Comment (kindermoumoute) [inline README.md]: Maybe put a graphic or link to external documentation?

File: README.md [hunk 5]
```diff
@@ -79,9 +106,9 @@ Creating a task manifest file.
                     "plugin_name": "statistics",
```
Comment (kindermoumoute) [inline README.md]: Line 96: I think "/intel/mock/*/baz" are string metrics. Will not work with the plugin. (maybe I'm wrong, you can test if the task work)

File: README.md [hunk 6]
```diff
@@ -36,21 +36,47 @@ $ make
...
+18) 9th-percentile - value below which 9% observations may be found 
+19) 25th-percentile - value below which 25% observations may be found 
+20) 75th-percentile - value below which 75% observations may be found 
+21) 91st-percentile - value below which 91% observations may be found  
+22) 95th-percentile - value below which 95% observations may be found 
+23) 98th-percentile - value below which 98% observations may be found 
+24) 99th-percenticle - value below which 99% observations may be found 
+
+To illustrate the sliding window concept, let's say we have a sliding window of 10 over 100 points long, if the sliding factor is 1 and we want to calculate the statistic "average", it will take the average from 0-10, 1-11, 2-12, 3-13 and so on. That takes the 100 data points and averages to 99 points.
+Sliding window can be used to smooth out a waveform thereby reducing noise. It is very useful in handling noisey data and also in live streaming of data.
```
Comment (kindermoumoute) [inline README.md]: 👍

File: README.md (hunk 1 above)
Comment (nanliu) [inline README.md]: plugin releases

File: README.md (hunk 2 above)
Comment (nanliu) [inline README.md]: './build' to indicate it's relative

File: README.md (hunk 3 above)
Comment (candysmurf) [inline README.md]: agreed.

File: README.md [hunk 7]
```diff
@@ -36,21 +36,47 @@ $ make
...
-
-Note: This Snap processor plugin changes the metric data type to `map[string]float64`. Any Snap publisher plugin used with this plugin should register the data type with gob. 
-
-```
-import "encoding/gob"
-
-gob.Register(map[string]float64{})
-```
+This Snap processor plugin calculates statistics over a sliding window. Currently, the plugin calculates the following statistics over a sliding window:
+1) Count - length of the buffer
```
Comment (candysmurf) [inline README.md]: let's not use the term - buffer.

File: README.md [hunk 8]
```diff
@@ -36,21 +36,47 @@ $ make
...
+2) Sum - total sum of data values in the buffer
```
Comment (candysmurf) [inline README.md]: The sum of data values in the sliding window?

File: README.md [hunk 9]
```diff
@@ -36,21 +36,47 @@ $ make
...
+3) Mean - average of data values 
+4) Median - middle data value in a sorted buffer
```
Comment (candysmurf) [inline README.md]: The middle data value in a sliding window.

File: README.md [hunk 10]
```diff
@@ -36,21 +36,47 @@ $ make
...
-import "encoding/gob"
-
-gob.Register(map[string]float64{})
-```
+This Snap processor plugin calculates statistics over a sliding window. Currently, the plugin calculates the following statistics over a sliding window:
+1) Count - length of the buffer
+2) Sum - total sum of data values in the buffer
+3) Mean - average of data values 
+4) Median - middle data value in a sorted buffer 
+5) Minimum - minimum of all data values
```
Comment (candysmurf) [inline README.md]: The minimum data value.

File: README.md [hunk 11]
```diff
@@ -36,21 +36,47 @@ $ make
...
-
-gob.Register(map[string]float64{})
-```
+This Snap processor plugin calculates statistics over a sliding window. Currently, the plugin calculates the following statistics over a sliding window:
+1) Count - length of the buffer
+2) Sum - total sum of data values in the buffer
+3) Mean - average of data values 
+4) Median - middle data value in a sorted buffer 
+5) Minimum - minimum of all data values  
+6) Maximum - maximum of all data values
```
Comment (candysmurf) [inline README.md]: the maximum value.

File: README.md [hunk 12]
```diff
@@ -36,21 +36,47 @@ $ make
...
-gob.Register(map[string]float64{})
-```
+This Snap processor plugin calculates statistics over a sliding window. Currently, the plugin calculates the following statistics over a sliding window:
+1) Count - length of the buffer
+2) Sum - total sum of data values in the buffer
+3) Mean - average of data values 
+4) Median - middle data value in a sorted buffer 
+5) Minimum - minimum of all data values  
+6) Maximum - maximum of all data values 
+7) Range - difference of largerst and smallest values
```
Comment (candysmurf) [inline README.md]: should you use max and min to be consistent.

File: README.md [hunk 13]
```diff
@@ -36,21 +36,47 @@ $ make
...
-```
+This Snap processor plugin calculates statistics over a sliding window. Currently, the plugin calculates the following statistics over a sliding window:
+1) Count - length of the buffer
+2) Sum - total sum of data values in the buffer
+3) Mean - average of data values 
+4) Median - middle data value in a sorted buffer 
+5) Minimum - minimum of all data values  
+6) Maximum - maximum of all data values 
+7) Range - difference of largerst and smallest values 
+8) Standard deviation - measure that quantifies the amount of varitation of a dataset
```
Comment (candysmurf) [inline README.md]: Probably be specific how it's calculated.

File: README.md [hunk 14]
```diff
@@ -36,21 +36,47 @@ $ make
...
+This Snap processor plugin calculates statistics over a sliding window. Currently, the plugin calculates the following statistics over a sliding window:
+1) Count - length of the buffer
+2) Sum - total sum of data values in the buffer
+3) Mean - average of data values 
+4) Median - middle data value in a sorted buffer 
+5) Minimum - minimum of all data values  
+6) Maximum - maximum of all data values 
+7) Range - difference of largerst and smallest values 
+8) Standard deviation - measure that quantifies the amount of varitation of a dataset 
+9) Variance - measures how far a set of numbers are spread out from their mean
```
Comment (candysmurf) [inline README.md]: give an example so that we know the calculation rule.

File: README.md [hunk 15]
```diff
@@ -36,21 +36,47 @@ $ make
...
+10) Mode - the number which appears most often in a dataset
+11) Kurtosis - measures the degree of peakedness of a distribution
```
Comment (candysmurf) [inline README.md]: give more specific explanation so that users know how this is calculated without reading code.

File: README.md [hunk 16]
```diff
@@ -36,21 +36,47 @@ $ make
...
+12) Skewness - measure of a dataset's symmetry or lack of symmetry
+13) Trimean - measure of a probability distribution's location as a weighted average of the distribution's median and its two quartiles
```
Comment (candysmurf) [inline README.md]: too abstract. Adding an example.

File: README.md [hunk 17]
```diff
@@ -36,21 +36,47 @@ $ make
...
+5) Minimum - minimum of all data values  
+6) Maximum - maximum of all data values 
+7) Range - difference of largerst and smallest values 
+8) Standard deviation - measure that quantifies the amount of varitation of a dataset 
+9) Variance - measures how far a set of numbers are spread out from their mean
+10) Mode - the number which appears most often in a dataset
+11) Kurtosis - measures the degree of peakedness of a distribution
+12) Skewness - measure of a dataset's symmetry or lack of symmetry
+13) Trimean - measure of a probability distribution's location as a weighted average of the distribution's median and its two quartiles 
+14) First quartile - middle number between the smallest number and the median of the dataset
```
Comment (candysmurf) [inline README.md]: Should this be consistent with "third quartile" explanation?

File: README.md [hunk 18]
```diff
@@ -36,21 +36,47 @@ $ make
...
+15) Third quartile - median of the upper half of the dataset
+16) Quartile range - measure of variability based on dividing a dataset into quartiles
```
Comment (candysmurf) [inline README.md]: please rewrite it so it's easy to understand

File: README.md [hunk 19]
```diff
@@ -36,21 +36,47 @@ $ make
...
+17) 2nd-percentile - value below which 2% observations may be found 
+18) 9th-percentile - value below which 9% observations may be found 
+19) 25th-percentile - value below which 25% observations may be found 
+20) 75th-percentile - value below which 75% observations may be found 
+21) 91st-percentile - value below which 91% observations may be found  
+22) 95th-percentile - value below which 95% observations may be found 
+23) 98th-percentile - value below which 98% observations may be found 
+24) 99th-percenticle - value below which 99% observations may be found
```
Comment (candysmurf) [inline README.md]: All these percentile definitions are hard to understand. Rewriting them. Aren't they observations at percentile?

File: README.md (hunk 4 above)
Comment (candysmurf) [inline README.md]: draw a diagram.

File: README.md [hunk 20]
```diff
@@ -59,8 +85,9 @@ Creating a task manifest file.
     "version": 1,
     "schedule": {
         "type": "simple",
-        "interval": "1s"
+        "interval": "100s"
```
Comment (candysmurf) [inline README.md]: if no sliding factor defined, what's the default?

File: README.md [hunk 21]
```diff
@@ -79,9 +106,9 @@ Creating a task manifest file.
                     "plugin_name": "statistics",
 		    "config":
     			{
-	    			"SlidingWindowLength": 15
+	    			"slidingWindowLength": 5,
+                    "slidingFactor": 2,
```
Comment (candysmurf) [inline README.md]: can you elaborate what's recommended "slidingwindorlength" and "slidingfactor" in terms of the data set size and the reasonings behind your recommendation.

File: README.md (hunk 20 above)
Comment (kindermoumoute) [inline README.md]: Sliding factor is 1 by default. So it's the interval.

Maybe 100 seconds is a bit too much for a task example?

File: README.md [hunk 22]
```diff
@@ -17,6 +17,7 @@ Snap plugin intended to process data and return statistics over a sliding window
 ## Getting Started
 ### System Requirements
 * Linux/amd64
+* OS X
```
Comment (candysmurf) [inline README.md]: MacOS

**Pull request merged**